### PR TITLE
Fixing `__iter__` returning private `cached_property` info

### LINKED ISSUE #7499: Private cached properties are included in BaseModel.__iter__
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Attributes whose name has a leading underscore are not treated as fields by Pydantic, and are not included in the model schema. However this doesn't seem to be the case for fields injected by `cached_property`.

### Example Code

```Python
from functools import cached_property

from pydantic import BaseModel


class Model(BaseModel):
    foo: int

    @cached_property
    def _foo(self):
        return -self.foo


m = Model(foo=1)
assert list(k for k, v in m) == ["foo"]

m._foo
keys = list(k for k, v in m)
assert keys == ["foo"], keys
```

Raises `AssertionError: ['foo', '_foo']`

### Python, Pydantic & OS Version

```Text
pydantic version: 2.3.0
        pydantic-core version: 2.6.3
          pydantic-core build: profile=release pgo=false
               python version: 3.10.11 | packaged by conda-forge | (main, May 10 2023, 19:01:19) [Clang 14.0.6 ]
                     platform: macOS-13.2.1-arm64-arm-64bit
     optional deps. installed: ['email-validator', 'typing-extensions']
```

### PR DESCRIPTION
**Pull request by sydney-runkle**
## Change Summary
* Fixing `__iter__` returning private `cached_property` info
* By default, private `computed_field`s have `repr` set to False

Fix #7499

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @samuelcolvin

### CODE REVIEW

File: pydantic/fields.py [hunk 1]
```diff
@@ -991,14 +992,26 @@ def computed_field(__func: PropertyT) -> PropertyT:
...
+
 def computed_field(
     __f: PropertyT | None = None,
     *,
     alias: str | None = None,
     alias_priority: int | None = None,
     title: str | None = None,
     description: str | None = None,
-    repr: bool = True,
+    repr: bool | None = None,
```
Comment (hramezani) [inline pydantic/fields.py:1014]: I think this would be a breaking change

Comment (sydney-runkle) [reply pydantic/fields.py:1014]: I still account for setting the default -> `True` within the body of the function if the field is private, so I don't think it's breaking.

File: pydantic/fields.py [hunk 2]
```diff
@@ -7,6 +7,7 @@
 import typing
 from copy import copy
 from dataclasses import Field as DataclassField
+from functools import cached_property
```
Comment (sydney-runkle) [inline pydantic/fields.py:10]: Still need to add checks for import errors (doesn't work in Python 3.7)

**Comment by sydney-runkle** [comment]
please review

File: pydantic/fields.py (hunk 1 above)
Comment (adriangb) [reply pydantic/fields.py:1014]: Do you mean if the field is public?

File: pydantic/fields.py (hunk 2 above)
Comment (adriangb) [reply pydantic/fields.py:10]: You could move the import to around line 1001

File: pydantic/main.py [hunk 3]
```diff
@@ -887,7 +887,7 @@ class MyModel(BaseModel, extra='allow'):
 
     def __iter__(self) -> TupleGenerator:
         """So `dict(model)` works."""
-        yield from self.__dict__.items()
+        yield from [(k, v) for (k, v) in self.__dict__.items() if not k.startswith('_')]
```
Comment (adriangb) [inline pydantic/main.py:890]: This seems like it should be it's own PR. The repr thing is one thing, the __iter__ change another

File: pydantic/fields.py (hunk 1 above)
Comment (hramezani) [reply pydantic/fields.py:1014]: I think this is a change for private fields. as the existing behavior seems to be a bug, so, it's good.
ignore my comment.

**Review by adriangb — APPROVED** [review]

File: pydantic/fields.py (hunk 1 above)
Comment (sydney-runkle) [reply pydantic/fields.py:1019]: @adriangb yep, I mean't public
